### PR TITLE
Using multiple discrete functions with TMOP for adaptivity

### PR DESCRIPTION
**Pull request by kmittal2**
This PR will allow us to specify multiple discrete functions for r-adaptivity with TMOP. 

Below is an example where we start with an indicator function and use it to derive discrete functions for size and aspect-ratio. The optimized mesh is obtained using the FD-based approach for TMOP.
image

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1352](https://github.com/mfem/mfem/pull/1352) | @kmittal2 | @v-dobrev | @vladotomov + @v-dobrev | 04/06/20 | 05/19/20 | 05/25/20 | |

### CODE REVIEW

**Comment by kmittal2** [comment]
LGTM

File: fem/tmop.hpp [hunk 1]
```diff
@@ -708,30 +708,59 @@ class DiscreteAdaptTC : public TargetConstructor
...
 
-   virtual ~DiscreteAdaptTC() { delete adapt_eval; }
+   virtual ~DiscreteAdaptTC()
+   {
+      delete adapt_eval;
+      delete tspec_fes;
+      delete tspec_fesv;
+   }
 
-   virtual void SetSerialDiscreteTargetSpec(GridFunction &tspec_);
```
Comment (vladotomov) [inline fem/tmop.hpp:732]: Lets keep this for now to have existing codes compiling. Inside we can just call the TargetSize version.

Comment (vladotomov) [reply fem/tmop.hpp:732]: Same for the Par version.

File: fem/tmop.cpp [hunk 2]
```diff
@@ -957,40 +957,169 @@ void AnalyticAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
...
+}
+
+void DiscreteAdaptTC::SetParDiscreteTargetBase(ParGridFunction &tspec_)
+{
+   const int vdim = tspec_.FESpace()->GetVDim(),
+             cnt = tspec_.Size()/vdim;
+
+   ncomp += vdim;
+
+   if (ncomp == vdim)
```
Comment (vladotomov) [inline fem/tmop.cpp:985]: Isn't `tspec_fes==NULL` a more intuitive check here?

File: fem/tmop_tools.hpp [hunk 3]
```diff
@@ -27,9 +27,10 @@ class AdvectorCG : public AdaptivityEvaluator
    RK4Solver ode_solver;
    Vector nodes0;
    Vector field0;
-
    const double dt_scale;
 
+   virtual void ComputeAtNewPositionScalar(const Vector &new_nodes,
```
Comment (vladotomov) [inline fem/tmop_tools.hpp:32]: No need to be virtual for now.

File: fem/tmop_tools.cpp [hunk 4]
```diff
@@ -66,13 +98,20 @@ void AdvectorCG::ComputeAtNewPosition(const Vector &new_nodes,
    {
       h_min = std::min(h_min, m->GetElementSize(i));
    }
-   double v_max = 0.0;
-   const int s = u.FESpace()->GetVSize() / 2;
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:70]: Huh was this assuming 2D?

File: fem/tmop_tools.cpp [hunk 5]
```diff
@@ -95,30 +140,34 @@ void AdvectorCG::ComputeAtNewPosition(const Vector &new_nodes,
...
-   const double minv = field0.Min(), maxv = field0.Max();
-   for (int i = 0; i < new_field.Size(); i++)
+   double glob_minv = minv;
+   double glob_maxv = maxv;
+#ifdef MFEM_USE_MPI
+   if (pfes)
    {
-      if (new_field(i) < minv) { new_field(i) = minv; }
-      if (new_field(i) > maxv) { new_field(i) = maxv; }
+      MPI_Allreduce(&minv, &glob_minv, 1, MPI_DOUBLE, MPI_MIN, pfes->GetComm());
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:154]: Yes this must be reduced globally, good catch.

File: fem/tmop_tools.cpp [hunk 6]
```diff
@@ -381,7 +430,7 @@ double TMOPNewtonSolver::ComputeScalingFactor(const Vector &x,
       if (energy_out > 1.2*energy_in || std::isnan(energy_out) != 0)
       {
          if (print_level >= 0)
-         { mfem::out << "Scale = " << scale << " Increasing energy.\n"; }
+         { mfem::out << "Scale = " << scale << " " << energy_out << " " << "Increasing energy.\n"; }
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:433]: Were you debugging something or actually changing the output? Indicate that the second number is the energy / the line is too long.

File: fem/tmop_tools.cpp [hunk 7]
```diff
@@ -29,13 +29,28 @@ void AdvectorCG::SetInitialField(const Vector &init_nodes,
...
+   new_field = field0;
+
+   for (int i = 0; i < ncomp; i++)
+   {
+      Vector new_field_temp(new_field.GetData()+i*pnt_cnt, pnt_cnt);
+      ComputeAtNewPositionScalar(new_nodes, new_field_temp);
+   }
 
+   //field0 = new_field;
+   //nodes0 = new_nodes;
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:46]: Why are these commented out? That would change the previous result significantly.

File: fem/tmop_tools.cpp [hunk 8]
```diff
@@ -29,13 +29,28 @@ void AdvectorCG::SetInitialField(const Vector &init_nodes,
 void AdvectorCG::ComputeAtNewPosition(const Vector &new_nodes,
                                       Vector &new_field)
 {
-#if defined(MFEM_DEBUG) || defined(MFEM_USE_MPI)
-   int myid = 0;
-#endif
-   Mesh *m = mesh;
+   // This function will not work for AMR meshes in the current state.
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:32]: Mark with TODO.

File: fem/tmop_tools.cpp [hunk 9]
```diff
@@ -29,13 +29,28 @@ void AdvectorCG::SetInitialField(const Vector &init_nodes,
...
+   const int dim     = fes->GetFE(0)->GetDim(),
+             ncomp   = fes->GetVDim(),
+             pnt_cnt = new_nodes.Size()/dim;
+
+   new_field = field0;
+
+   for (int i = 0; i < ncomp; i++)
+   {
+      Vector new_field_temp(new_field.GetData()+i*pnt_cnt, pnt_cnt);
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:41]: This assumes that the #dofs in the components `new_field` equals the number of nodes in the mesh. Shouldn't the size be `new_field.Size() / ncomp`?

File: fem/tmop_tools.cpp (hunk 4 above)
Comment (kmittal2) [reply fem/tmop_tools.cpp:70]: yep

File: fem/tmop_tools.cpp (hunk 6 above)
Comment (kmittal2) [reply fem/tmop_tools.cpp:428]: I think this is from debugging. I will undo this change.

File: fem/tmop.cpp [hunk 10]
```diff
@@ -1163,6 +1444,7 @@ void AdaptivityEvaluator::SetParMetaInfo(const ParMesh &m,
    delete pmesh;
    pmesh = new ParMesh(m, true);
    pfes  = new ParFiniteElementSpace(pmesh, &fec, num_comp);
+   fes   = pfes;
```
Comment (vladotomov) [inline fem/tmop.cpp:1447]: Why is this needed?

File: fem/tmop.cpp [hunk 11]
```diff
@@ -957,42 +957,181 @@ void AnalyticAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
...
+
+   for (int i = 0; i < cnt*vdim; i++)
+   {
+      tspec(i+(ncomp-vdim)*cnt) = tspec_(i);
+   }
+}
+
+void DiscreteAdaptTC::SetParDiscreteTargetSize(ParGridFunction &tspec_)
+{
+   MFEM_VERIFY(sizeidx == -1, " Size discrete function already specified")
```
Comment (vladotomov) [inline fem/tmop.cpp:1013]:
```suggestion
   if (sizeidx > -1) { SetTspecAtIndex(sizeidx, stpec_); }
```
This would allow to call these functions after Finalize() to update the targets.

File: fem/tmop.cpp [hunk 12]
```diff
@@ -1040,53 +1187,190 @@ void DiscreteAdaptTC::ComputeElementTargets(int e_id, const FiniteElement &fe,
...
-   if (tspec_perth.Size() != x.Size())
-   {
-      tspec_perth.SetSize(x.Size());
-   }
+   tspec_perth.SetSize(x.Size()*ncomp);
 
-   Vector TSpecTemp;
-   Vector xtemp = x;
+   Vector TSpecTemp(ncomp*cnt);
+   Vector xtemp(x.GetData(), x.Size());
```
Comment (vladotomov) [inline fem/tmop.cpp:1368]: It might be slower, but I'd prefer to do a copy here. Otherwise we implicitly put the assumption that the rest of the code in this function will not change `xtemp` (or change it and revert it), which is true for now, but might lead to bugs in the future.

File: fem/tmop.hpp [hunk 13]
```diff
@@ -726,20 +736,57 @@ class DiscreteAdaptTC : public TargetConstructor
...
+       call a Finalize method.
+
+       @param[in] tspec_  Input values of a geometric parameter. Note that
+                          the methods in this class support only functions that
+                          use H1_FECollection collection of the same order. */
+   ///@{
+   virtual void SetSerialDiscreteTargetSpec(const GridFunction &tspec_);
+   virtual void SetSerialDiscreteTargetSize(const GridFunction &tspec_);
+   virtual void SetSerialDiscreteTargetSkew(const GridFunction &tspec_);
+   virtual void SetSerialDiscreteTargetAspectRatio(const GridFunction &tspec_);
```
Comment (vladotomov) [inline fem/tmop.hpp:778]: Could you add a short comment how many components are expected for a given dimension, and what the components represent in the skew / aspect ratio / orientation cases.

Comment (vladotomov) [reply fem/tmop.hpp:778]: Maybe is should be done in the general block comment above.

**Comment by v-dobrev** [comment]
@kmittal2, can you resolve the conflicts by merging `master` here?

Also, it will be good to mention this new capability in the `CHANGELOG`.

**Comment by kmittal2** [comment]
> @kmittal2, can you resolve the conflicts by merging `master` here?
> 
> Also, it will be good to mention this new capability in the `CHANGELOG`.

Done.

**Comment by v-dobrev** [comment]
Merged in `next` for testing...
